### PR TITLE
App/Client: fixes + features + unit-tests + refactoring + documentation

### LINKED ISSUE #356: Improve server tunnel access list implementation
**By submitting this issue, I confirm the following:**
- I have read and understood the [contributor guide](https://github.com/monero-project/kovri/blob/master/doc/CONTRIBUTING.md).
- I have checked that the issue I am reporting can be replicated or that the feature I am suggesting is not present.
- I have checked opened or recently closed [pull requests](https://github.com/monero-project/kovri/pulls) for existing solutions/implementations to my issue/suggestion.

_Place an X inside the bracket to confirm_
- [X] I confirm.

---

I've spent very little time with streaming and client handling of _server_ tunnels. Also, I've also never known server tunnels to work - so this is only being ticketed now.

Router is well-integrated, 30+ minutes; built against 00e0dfe6ce1b458cf28bd453b858b1f57fbe4200:

`tunnels.conf`:

```
[LOCALSITE]
type = server
host = 127.0.0.1
port = 80
keys = site-keys.dat
inport = 8080
```

``` bash
# ./kovri --log-to-console 0  # using default log-levels
$ tail -f ~/.kovri/logs/* | grep -E "Stream|Garlic|ClientDest|NetDb|Lease"
```

Connect from a java i2p configured browser:

```
DBG   I2NPMessage: Garlic
DBG   GarlicDestination: 1 cloves
DBG   GarlicDestination: Garlic type local
NFO   GarlicDestination: message 1075046620 acknowledged
WRN   ClientDestination: publish confirmation was not received in 5seconds. Try again
DBG   ClientDestination: publish LeaseSet of styoaws7kp5ceeqtw2kqhviomadyrt6aoqza5jrkcm4sarrypw4q
WRN   GarlicRoutingSession: no garlic tags available, using ElGamal
DBG   I2NPMessage: Garlic
DBG   GarlicDestination: 1 cloves
DBG   GarlicDestination: Garlic type local
NFO   GarlicDestination: message 348594653 acknowledged
DBG   I2NPMessage: Garlic
DBG   GarlicDestination: 3 cloves
DBG   GarlicDestination: Garlic type tunnel
DBG   GarlicDestination: Garlic type local
DBG   ClientDestination: remote LeaseSet
DBG   LeaseSet: num=3
NFO   LeaseSet: lease's tunnel gateway not found, requesting
DBG   ClientDestination: remote LeaseSet updated
DBG   GarlicDestination: Garlic type destination
DBG   Stream: received seqn=0
DBG   Stream: process seqn=0, flags=1257
DBG   Stream: synchronize
NFO   Stream: from identity 2urKDqBEn6MJ5POf8TG2omkO8QKd6zWHxryHrGlxmRw=
DBG   Stream: incoming stream from 2urKDqBEn6MJ5POf8TG2omkO8QKd6zWHxryHrGlxmRw=
DBG   Stream: max packet size 1730
DBG   Stream: signature
NFO   Stream: trying to send stream data before closing
DBG   NetDb: DatabaseStore
DBG   NetDb: RouterInfo
DBG   I2NPMessage: Garlic
DBG   GarlicDestination: 1 cloves
DBG   GarlicDestination: Garlic type local
NFO   GarlicRoutingSession: leaseset update confirmed
NFO   GarlicDestination: message 849121881 acknowledged
DBG   I2NPMessage: Garlic
DBG   GarlicDestination: 3 cloves
DBG   GarlicDestination: Garlic type tunnel
DBG   GarlicDestination: Garlic type local
DBG   ClientDestination: remote LeaseSet
DBG   LeaseSet: num=3
NFO   LeaseSet: lease's tunnel gateway not found, requesting
WRN   NetDb: destination -7~KOyL3JJmE96zH7t4aaoIHNz3f1DcMBdpFKHrdUIw= was already requested
DBG   ClientDestination: remote LeaseSet updated
DBG   GarlicDestination: Garlic type destination
DBG   Stream: packet 0 acknowledged rtt=2274
NFO   Stream: FIN sent
DBG   Stream: plain ACK received
DBG   Stream: stream deleted
WRN   ClientDestination: publish confirmation was not received in 5seconds. Try again
DBG   ClientDestination: publish LeaseSet of styoaws7kp5ceeqtw2kqhviomadyrt6aoqza5jrkcm4sarrypw4q
WRN   GarlicRoutingSession: no garlic tags available, using ElGamal
```

10+ minutes later, after many more attempts, the following is still reported:

```
DBG   I2NPMessage: Garlic
DBG   GarlicDestination: 1 cloves
DBG   GarlicDestination: Garlic type destination
DBG   Stream: received seqn=0
DBG   Stream: process seqn=0, flags=1257
DBG   Stream: synchronize
NFO   Stream: from identity 2urKDqBEn6MJ5POf8TG2omkO8QKd6zWHxryHrGlxmRw=
DBG   Stream: incoming stream from 2urKDqBEn6MJ5POf8TG2omkO8QKd6zWHxryHrGlxmRw=
DBG   Stream: max packet size 1730
DBG   Stream: signature
NFO   Stream: trying to send stream data before closing
DBG   I2NPMessage: Garlic
DBG   GarlicDestination: 1 cloves
DBG   GarlicDestination: Garlic type local
NFO   GarlicDestination: message 1951065053 acknowledged
NFO   NetDb: exploring new 4 routers
DBG   I2NPMessage: Garlic
DBG   GarlicDestination: 1 cloves
DBG   GarlicDestination: Garlic type destination
DBG   Stream: packet 0 acknowledged rtt=2298
NFO   Stream: FIN sent
DBG   Stream: plain ACK received
DBG   Stream: stream deleted
DBG   I2NPMessage: Garlic
DBG   GarlicDestination: 1 cloves
DBG   GarlicDestination: Garlic type destination
WRN   StreamingDestination: unknown stream 2351358304
```

Inbound port `8080` never reached, browser reports `Could not find the following destination: http://styoaws7kp5ceeqtw2kqhviomadyrt6aoqza5jrkcm4sarrypw4q.b32.i2p`.

### LINKED ISSUE #486: Allow multiple tunnel destinations + custom ports in client tunnel
**By submitting this issue, I confirm the following:**

- I have read and understood the [contributor guide](https://github.com/monero-project/kovri/blob/master/doc/CONTRIBUTING.md).
- I have checked that the issue I am reporting can be replicated or that the feature I am suggesting is not present.
- I have checked opened or recently closed [pull requests](https://github.com/monero-project/kovri/pulls) for existing solutions/implementations to my issue/suggestion.

*Place an X inside the bracket to confirm*
- [X] I confirm.

---

I'm currently working in this area within my fork in branch `config` so this wouldn't take much work. This feature requires some starting work on #9 but will allow randomized choices for tunnel destination by a comma-separated value, e.g; `irc.dg.i2p,irc.postman.i2p,irc.echelon.i2p` (or similar).

### PR DESCRIPTION
**Pull request by anonimal**
**By submitting this pull-request, I confirm the following:**

- I have read and understood the [Contributing Guide](https://github.com/monero-project/kovri/blob/master/doc/CONTRIBUTING.md).
- I have checked that another pull-request for this purpose does not exist.
- I have considered and confirmed that this submission will be valuable to others.
- I accept that this submission may not be used and that this pull-request may be closed by the will of the maintainer.
- I give this submission freely under the BSD 3-clause license.

*Place an X inside the bracket (or click the box in preview) to confirm*
- [X] I confirm.

---

See git log entries for details.

There are various TODO's that I will resolve while this PR is open. Overall, the work is in progress and will continue (sooner or later) after merging (many in-depth TODO's all across the board).

Fixes #356. Resolves #486. References #340 #96 #34 #9

### CODE REVIEW

**Comment by anonimal** [comment]
JFTR: unit-tests and live tests (on linux) pass. I'll do testing on all platforms and also fix the win build before merging.

**Comment by anonimal** [comment]
**TL;DR:** the TODO's just keep piling up but there's nothing keeping this PR from being merged at this point.

Live tests pass on **all** regularly tested platforms with the exception of ARMv8 but the issue on that machine may be completely unrelated or indirectly related as a result of new fixes/features that exposed previously unknown problems. I don't know, I'll have to talk with @danrmiller to see if there was a system update which may have had an impact (the issue is a thrown boost `exception_ptr`, possibly related to cpp-netlib during subscription updating but I've touched nothing in that area. I'll need more time to test and since the ARMv8 box is maxed out resource-wise, I'll have to wait in line).

I don't think the ARMv8 machine should hold back this merge so I'd like to go ahead and merge now. My commits are getting too large (sorry about that) so merging with help me with atomicity and streamlining of TODO's while providing new fixes/features for others to use (hey, that rhymed!) :dog:

**Comment by anonimal** [comment]
Confirmed: the issue is reproducible against master - though - the last code changes in master were from November 26th and I had since ran an ARMv8 instance 24x7 for X days until recently (when I took the instance offline). I'll open a new unrelated-to-#487 ticket.
